math/inv_sqrt.cpp

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file
+ * @brief Calculate the inverse inverse root.
+ * @details
+ * Two implementation to calculate inverse inverse root,
+ * from Quake III Arena (C++ version) and with a standard library (`cmath`)
+ */
+
+#include <cmath>        /// for `std::sqrt`
+#include <iostream>     /// for IO operations
+#include <type_traits>  /// for `static_assert`
+
+/**
+ * @brief This is the function that calculates the fast inverse square root.
+ * The following code is the fast inverse square root implementation from
+ * Quake III Arena (Adapted for C++) More information can be found at
+ * [Wikipedia](https://en.wikipedia.org/wiki/Fast_inverse_square_root)
+ * @tparam T floating type
+ * @tparam iterations inverse square root, the greater the number of
+ * iterations, the more exact the result will be (1 or 2).
+ * @param x value to calculate
+ * @return T return inverse square root
+ */
+template <typename T, char iterations = 2>
+inline T Fast_InvSqrt(T x) {
+    using Tint = typename std::conditional<sizeof(T) == 8, std::int64_t,
+                                           std::int32_t>::type;
+    T y = x;
+    T x2 = y * 0.5;
+    Tint i = *(Tint *)&y;
+    i = (sizeof(T) == 8 ? 0x5fe6eb50c7b537a9 : 0x5f3759df) - (i >> 1);
+    y = *(T *)&i;
+
+    y = *reinterpret_cast<T *>(&i);
+
+    y = y * (1.5 - (x2 * y * y));
+    if (iterations == 2) {
+        y = y * (1.5 - (x2 * y * y));
+    }
+    return y;
+}
+
+/**
+ * @brief This is the function that calculates the fast inverse square root.
+ * The following code is the fast inverse square root with standard lib (cmath)
+ * More info:
+ * https://www.linkedin.com/pulse/fast-inverse-square-root-still-armin-kassemi-langroodi
+ * @tparam T floating type
+ * @param x value to calculate
+ * @return T return inverse square root
+ */
+template <typename T>
+T Standard_InvSqrt(T number) {
+    T squareRoot = sqrt(number);
+    return 1.0f / squareRoot;
+}
+
```

Comment:
Please add self-test implementations using `assert` (see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program as an example).

Reply:
Tests have been added